pytorch_lightning/plugins/training_type/deepspeed.py

```diff
@@ -0,0 +1,215 @@
+from pytorch_lightning.utilities.imports import _DEEPSPEED_AVAILABLE
+from pytorch_lightning.utilities.seed import seed_everything
+
+if _DEEPSPEED_AVAILABLE:
+    import deepspeed
+else:
+    deepspeed = None
+
+
+class LightningDeepSpeedModule(_LightningModuleWrapperBase):
+
+    def __init__(self, pl_module: LightningModule, precision: int):
+        super().__init__(pl_module)
+        self.module = pl_module
+        self.precision = precision
+
+    def forward(self, *inputs, **kwargs):
+        if self.precision == 16:
+            inputs = move_float_tensors_to_half(inputs)
+        return super().forward(*inputs, **kwargs)
+
+
+class DeepSpeedPlugin(DDPPlugin):
+    distributed_backend = "deepspeed"
+
+    def __init__(
+        self,
+        config: Union[Path, str, dict],
+        logging_level: int = logging.WARN,
+        num_nodes=1,
+        parallel_devices: List[torch.device] = None,
+        cluster_environment: ClusterEnvironment = None,
+    ) -> None:
+        super().__init__(
+            parallel_devices=parallel_devices, num_nodes=num_nodes, cluster_environment=cluster_environment
+        )
+        if isinstance(config, str) or isinstance(config, Path):
+            with open(config) as f:
+                self.config = json.load(f)
+        else:
+            self.config = config
+        self._config_initialized = False
+        deepspeed.utils.logging.logger.setLevel(logging_level)
+
+    def pre_training(self):
+        self.set_world_ranks()
+        self.init_ddp_connection(self.global_rank, self.world_size)
+
+        self.init_deepspeed()
+
+        # TODO: check if needed
+        seed = os.environ.get("PL_GLOBAL_SEED")
+        if seed is not None:
+            seed_everything(int(seed))
+
+        # set warning rank
+        rank_zero_only.rank = self.global_rank
+
+        # set the ranks and devices
+        self.dist.rank = self.global_rank
+        self.dist.device = self.root_device
+
+        # move the model to the correct device
+        self.model_to_device()
+        self.barrier()
+
+    def init_deepspeed(self):
+        if not self._config_initialized:
+            self._format_config()
+            self._config_initialized = True
+
+        precision = self.lightning_module.trainer.accelerator_backend.precision
+        model_parameters = filter(lambda p: p.requires_grad, self.model.parameters())
+        model, optimizer, _, lr_scheduler = deepspeed.initialize(
+            args=SimpleNamespace(local_rank=self.local_rank),
+            model=LightningDeepSpeedModule(pl_module=self.model, precision=precision),
+            model_parameters=model_parameters,
+            config_params=self.config,
+        )
+        trainer = self.lightning_module.trainer
+        if self.lightning_module.training:
+            trainer.optimizers = [optimizer]
+            trainer.lr_schedulers = self.configure_scheduler(lr_scheduler)
+            trainer.convert_to_lightning_optimizers()
```

Comment:
What do you mean automatically inferred? It has to be set manually as this happens outside config optimizers from the `deepspeed.initialize function`

Reply:
It isn't needed, as it will be wrap automacally if _lightning_optimizers is still None.
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/properties.py#L361